File: ModBoucWen/ModBoucWen.cpp
/* ****************************************************************** **
**    OpenSees - Open System for Earthquake Engineering Simulation    **
**          Pacific Earthquake Engineering Research Center            **
**                                                                    **
**                                                                    **
** (C) Copyright 1999, The Regents of the University of California    **
** All Rights Reserved.                                               **
**                                                                    **
** Commercial use of this program without express permission of the   **
** University of California, Berkeley, is strictly prohibited.  See   **
** file 'COPYRIGHT'  in main directory for information on usage and   **
** redistribution,  and for a DISCLAIMER OF ALL WARRANTIES.           **
**                                                                    **
** Developed by:                                                      **
**   Frank McKenna (fmckenna@ce.berkeley.edu)                         **
**   Gregory L. Fenves (fenves@ce.berkeley.edu)                       **
**   Filip C. Filippou (filippou@ce.berkeley.edu)                     **
**                                                                    **
** ****************************************************************** */
                                                                        
// Written: Wenchen Lie (666@e.gzhu.edu.cn)
// Created: July 26, 2024
// Revision: A
//
// Description: This file contains the implementation of the
// ModBoucWen class.


#include <ModBoucWen.h>
#include <Vector.h>
#include <Matrix.h>
#include <Channel.h>
#include <Information.h>
#include <Parameter.h>

#include <string.h>

#include <math.h>
#include <float.h>


#include <elementAPI.h>
#include <OPS_Globals.h>

static int numModBoucWenMaterials = 0;

void *
OPS_ModBoucWen()
{
  // Pointer to a uniaxial material that will be returned
  UniaxialMaterial *theMaterial = 0;

  int    iData[1];
  double dData[9];
  int    iter;
  int numData = 1;

  // Print information
  if (numModBoucWenMaterials == 0) {
      opserr << "Modified BoucWen unaxial material - Written by Wenchen Lie (July 27, 2024)\n";
      numModBoucWenMaterials++;
  }

  if (OPS_GetIntInput(&numData, iData) != 0) {
      opserr << "WARNING invalid uniaxialMaterial ModBoucWen tag" << endln;
      return 0;
  }

  numData = OPS_GetNumRemainingInputArgs();
  int num1 = 9;
  int num2 = 1;

  if (numData != 9 && numData != 10) {
      opserr << "Invalid args for uniaxialMaterial ModBoucWen " << iData[0] << " Fy? uy? alpha? n? Q? b? A? beta? gamma? <iter?>" << endln;
      return 0;
  }

  if (OPS_GetDoubleInput(&num1, dData) != 0) {
      opserr << "Invalid args for uniaxialMaterial ModBoucWen " << iData[0] << " Fy? uy? alpha? n? Q? b? A? beta? gamma? <iter?>" << endln;
      return 0;
  }

  if (numData == 9) {
      iter = MODBOUCWEN_DEFAULT_ITER;
  }
  else {
      if (OPS_GetIntInput(&num2, &iter) != 0) {
          opserr << "WARNING cannot get $iter, want int but got double" << endln;
          return 0;
      }
  }

  int tag = iData[0];
  double Fy, uy, alpha, n, Q, b, A, beta, gamma;
  Fy = dData[0];
  uy = dData[1];
  alpha = dData[2];
  n = dData[3];
  Q = dData[4];
  b = dData[5];
  A = dData[6];
  beta = dData[7];
  gamma = dData[8];

  if (Fy <= 0) {
      opserr << "WARNING Fy must be positive" << endln;
      return 0;
  }
  if (uy <= 0) {
      opserr << "WARNING uy must be positive" << endln;
      return 0;
  }
  if (alpha < 0) {
      opserr << "WARNING alpha must not less than 0" << endln;
      return 0;
  }
  if (n <= 0) {
      opserr << "WARNING n must be positive" << endln;
      return 0;
  }
  if (iter <= 0) {
      opserr << "WARNING iter must be positive" << endln;
      return 0;
  }
  //opserr << "Fy = " << Fy << "\n" << endln;
  //opserr << "uy = " << uy << "\n" << endln;
  //opserr << "alpha = " << alpha << "\n" << endln;
  //opserr << "n = " << n << "\n" << endln;
  //opserr << "Q = " << Q << "\n" << endln;
  //opserr << "b = " << b << "\n" << endln;
  //opserr << "beta = " << beta << "\n" << endln;
  //opserr << "gamma = " << gamma << "\n" << endln;
  //opserr << "iter = " << iter << "\n" << endln;


  // Parsing was successful, allocate the material
  theMaterial = new ModBoucWen(tag, Fy, uy, alpha, n, Q, b, A, beta, gamma, iter);
  if (theMaterial == 0) {
    opserr << "WARNING could not create uniaxialMaterial of type ModBoucWen Material" << endln;
    return 0;
  }

  return theMaterial;
}



ModBoucWen::ModBoucWen(int tag_, double Fy_, double uy_, double alpha_, double n_,double Q_, double b_, double A_, double beta_, double gamma_,int iter_)
    :UniaxialMaterial(tag_, MAT_TAG_ModBoucWen),
    Fy(Fy_), uy(uy_), alpha(alpha_), n(n_), Q(Q_), b(b_), A(A_), beta(beta_), gamma(gamma_), iter(iter_)
{
    Cstrain = 0.0;
    Cstress = 0.0;
    Tstrain = 0.0;
    Tstress = 0.0;
    Ttangent = Fy_ / uy_;
    Ctangent = Fy_ / uy_;
    Tz = 0.0;
    Cz = 0.0;
    Twp = 0.0;
    Cwp = 0.0;
    Tface = uy;
    Cface = uy;
}

ModBoucWen::ModBoucWen():UniaxialMaterial(0, MAT_TAG_ModBoucWen),
    Fy(0.0), uy(0.0), alpha(0.0), n(0.0), Q(0.0), b(0.0), A(0.0), beta(0.0), gamma(0.0), iter(0)
{

}

ModBoucWen::~ModBoucWen ()
{
    
}

int ModBoucWen::setTrialStrain (double strain, double strainRate)
{
    Tstrain = Cstrain;
    Tstress = Cstress;
    Ttangent = Ctangent;
    Tz = Cz;
    Twp = Cwp;
    Tface = Cface;

   // Determine change in strain from last converged state
   double dStrain = strain - Cstrain;

    if (dStrain == 0.0) {
        return 0;
    }
     // Set trial strain
    Tstrain = strain;
    double dStrain_ = dStrain / iter;
    double z_ = Cz;
    double strain_;
    double sgn;
    double m;
    double S1, S2, S3, S4;
    for (int i = 0; i < iter; i++) {
        // substep
        strain_ = Cstrain + dStrain_ * i;
        if (strain_ > Tface) {
            // positive yielding
            Twp = Twp + strain_ - Tface;
            Tface = strain_;
        }
        else if (strain_ < Tface - 2 * uy) {
            // negative yielding
            Twp = Twp + Tface - 2 * uy - strain_;
            Tface = strain_ + 2 * uy;
        }
        if (dStrain_ * z_ < 0.0) {
            sgn = -1.0;
        }
        else if (dStrain_ * z_ == 0.0) {
            sgn = 0.0;
        }
        else {
            sgn = 1.0;
        }
        m = 1 + Q * (1 - pow(b, -Twp / uy));
        S1 = 1.0 / uy * (A - (beta * sgn + gamma) * pow(fabs(z_ / m), n));
        S2 = 1.0 / uy * (A - (beta * sgn + gamma) * pow(fabs(z_ / m + 0.5 * dStrain_ * S1), n));
        S3 = 1.0 / uy * (A - (beta * sgn + gamma) * pow(fabs(z_ / m + 0.5 * dStrain_ * S2), n));
        S4 = 1.0 / uy * (A - (beta * sgn + gamma) * pow(fabs(z_ / m + dStrain_ * S3), n));
        z_ = z_ + 1.0 / 6.0 * dStrain_ * (S1 + S2 + S3 + S4);
        Tstress = alpha * Fy / uy * strain_ + (1 - alpha) * Fy * z_;
    };
    Tz = z_;
    Ttangent = (Tstress - Cstress) / dStrain;
}

double ModBoucWen::getStrain ()
{
   return Tstrain;
}

double ModBoucWen::getStress ()
{
   return Tstress;
}

double ModBoucWen::getTangent ()
{
    return Ttangent;
}

int ModBoucWen::commitState ()
{
   Cstrain = Tstrain;
   Cstress = Tstress;
   Ctangent = Ttangent;
   Cz = Tz;
   Cwp = Twp;
   Cface = Tface;
   return 0;
}

int ModBoucWen::revertToLastCommit ()
{
   Tstrain = Cstrain;
   Tstress = Cstress;
   Ttangent = Ttangent;
   Tz = Cz;
   Twp = Cwp;
   Tface = Cface;
   return 0;
}

int ModBoucWen::revertToStart ()
{
   Cstrain = 0.0;
   Cstress = 0.0;
   Ttangent = Fy / uy;
   Ctangent = Fy / uy;
   Tstrain = 0.0;
   Tstress = 0.0;
   Tz = 0.0;
   Twp = 0.0;
   Tface = uy;
   Cz = 0.0;
   Cwp = 0.0;
   Cface = uy;
   return 0;
}

UniaxialMaterial* ModBoucWen::getCopy ()
{
   ModBoucWen* theCopy = new ModBoucWen(this->getTag(), Fy, uy, alpha, n, Q, b, A, beta, gamma, iter);
   return theCopy;
}

int ModBoucWen::sendSelf (int commitTag, Channel& theChannel)
{
    opserr << "Currently ModBoucWen::sendSelf() is not avaiable for ModBoucWen material" << endln;
    return -1;
}

int ModBoucWen::recvSelf (int commitTag, Channel& theChannel,
                                FEM_ObjectBroker& theBroker)
{
    opserr << "Currently ModBoucWen::recvSelf() is not avaiable for ModBoucWen material" << endln;
    return -1;

}

void ModBoucWen::Print (OPS_Stream& s, int flag)
{
  if (flag == OPS_PRINT_PRINTMODEL_MATERIAL) {    
    s << "ModBoucWen tag:   " << this->getTag() << endln;
    s << "  Fy:         " << Fy << " ";
    s << "  uy:         " << uy << " ";
    s << "  alpha:      " << alpha << " ";
    s << "  n:          " << n << " ";
    s << "  Q:          " << Q << " ";
    s << "  b:          " << b << " ";
    s << "  A:          " << A << " ";
    s << "  beta:       " << beta << " ";
    s << "  gamma:      " << gamma << " ";
    s << "  iter:       " << iter << " ";
  }

  if (flag == OPS_PRINT_PRINTMODEL_JSON) {
    s << "\t\t\t{";
	s << "\"tag\": \"" << this->getTag() << "\", ";
    s << "\"type\": \"ModBoucWen\", ";
    s << "\"Fy\": " << Fy << ", ";
    s << "\"uy\": " << uy << ", ";
    s << "\"alpha\": " << alpha << ", ";
	s << "\"n " << n << ", ";
    s << "\"Q\": " << Q << ", ";
    s << "\"b\": " << b << ", ";
    s << "\"A\": " << A << ", ";
    s << "\"beta\": " << beta << ", ";
    s << "\"gamma\": " << gamma << ", ";
    s << "\"iter\": " << iter << ", ";
  }
  
}

File: src/ModBoucWen/ModBoucWen.py
import math
import sys
import numpy as np
import matplotlib.pyplot as plt
from ..UniaxialMaterial import UniaxialMaterial


class ModBoucWen(UniaxialMaterial):
    
    def __init__(self,
            tag: int,
            Fy: float,
            uy: float,
            alpha: float,
            n: float,
            Q: float,
            b: float,
            A: float,
            beta: float,
            gamma: float,
            iter: int=10
        ):
        self.tag = tag
        # Materail parameters
        self.Fy = Fy
        self.uy = uy
        self.alpha = alpha
        self.n = n
        self.Q = Q
        self.b = b
        self.A = A
        self.beta = beta
        self.gamma = gamma
        self.iter = iter
        # Response history
        self.Cstrain: float  # 上一步的应变
        self.Tstrain: float  # 当前步的应变
        self.Cstress: float  # 上一步的应力
        self.Tstress: float # 当前步的应力
        self.Ctangent: float  # 上一步的切线刚度
        self.Ttangent: float  # 当前步的切线刚度
        self.Cz: float  # 滞回变量z
        self.Tz: float
        self.Cwp: float  # 累积塑性应变
        self.Twp: float
        self.Cface: float  # 正向屈服面
        self.Tface: float
        self._check_paras()
        self._init_paras()

    def _check_paras(self):
        assert self.Fy > 0, "Fy must be positive"
        assert self.uy > 0, "uy must be positive"
        assert self.alpha >= 0, "alpha must not less than 0"
        assert self.n > 0, "n must be positive"

    def _init_paras(self):
        self.k0 = self.Fy / self.uy
        self.Cstrain = 0
        self.Tstrain = 0
        self.Cstress = 0
        self.Tstress = 0
        self.Ctangent = self.k0
        self.Ttangent = self.k0
        self.Cz = 0
        self.Tz = 0
        self.Cwp = 0
        self.Twp = 0
        self.Cface = self.uy
        self.Tface = self.uy

    def setTrialStrain(self, strain, strainRate=0):
        """传入当前步的应变值strain"""
        # Reset history variables to last converged state
        self.Tstrain = self.Cstrain
        self.Tstress = self.Cstress
        self.Ttangent = self.Ctangent
        self.Tz = self.Cz
        self.Twp = self.Cwp
        self.Tface = self.Cface
        dStrain = strain - self.Cstrain
        if dStrain == 0:
            return
        else:
            self.Tstrain = strain
        # Calculate stress and tangent
        dStrain_ = dStrain / self.iter
        z_ = self.Cz  # 临时的z值
        for i in range(self.iter):
            # 迭代iter个子步
            strain_ = self.Cstrain + dStrain_ * i  # 每一步应变
            if strain_ > self.Tface:
                # 正向屈服
                self.Twp += strain_ - self.Tface
                self.Tface = strain_
            elif strain_ < self.Tface - 2 * self.uy:
                # 负向屈服
                self.Twp += self.Tface - 2 * self.uy - strain_
                self.Tface = strain_ + 2 * self.uy
            if dStrain_ * z_ < 0.0:
                sgn = -1.0
            elif dStrain_ * z_ == 0.0:
                sgn = 0.0
            else:
                sgn = 1.0
            m = 1 + self.Q * (1 - pow(self.b, -self.Twp / self.uy))
            S1 = 1.0 / self.uy * (self.A - (self.beta * sgn + self.gamma) * pow(abs(z_ / m), self.n))
            S2 = 1.0 / self.uy * (self.A - (self.beta * sgn + self.gamma) * pow(abs(z_ / m + 0.5 * dStrain_ * S1), self.n))
            S3 = 1.0 / self.uy * (self.A - (self.beta * sgn + self.gamma) * pow(abs(z_ / m + 0.5 * dStrain_ * S2), self.n))
            S4 = 1.0 / self.uy * (self.A - (self.beta * sgn + self.gamma) * pow(abs(z_ / m + dStrain_ * S3), self.n))
            z_ = z_ + 1.0 / 6.0 * dStrain_ * (S1 + S2 + S3 + S4)
            self.Tstress = self.alpha * self.Fy / self.uy * strain_ + (1 - self.alpha) * self.Fy * z_
        else:
            # 将临时变量赋给实例变量
            self.Tz = z_
        self.Ttangent = (self.Tstress - self.Cstress) / dStrain

    def commitState(self):
        self.Cstrain = self.Tstrain
        self.Cstress = self.Tstress
        self.Ctangent = self.Ttangent
        self.Cz = self.Tz
        self.Cwp = self.Twp
        self.Cface = self.Tface

    def getStrain(self):
        return self.Tstrain

    def getStress(self):
        return self.Tstress

    def getTangent(self):
        return self.Ttangent
    

